tests/build/_internal/config/test_assemble_build_config.py

```diff
@@ -96,11 +102,60 @@ def test_assembles_config_using_all_relevant_files(self):
             create_files(directory, [unused_mbed_lib_file])
 
             subject = _assemble_config_from_sources_and_lib_files(
-                target_source, find_files("mbed_lib.json", Path(directory)), created_mbed_app_file
+                target, find_files("mbed_lib.json", Path(directory)), created_mbed_app_file
             )
 
             mbed_lib_sources = [Source.from_mbed_lib(Path(directory, file), ["A"]) for file in created_mbed_lib_files]
             mbed_app_source = Source.from_mbed_app(created_mbed_app_file, ["A"])
-            expected_config = Config.from_sources([target_source] + mbed_lib_sources + [mbed_app_source])
+            expected_config = Config.from_sources([Source.from_target(target)] + mbed_lib_sources + [mbed_app_source])
 
             self.assertEqual(subject, expected_config)
+
+    def test_updates_target_labels_from_config(self):
+        target = {
+            "config": {"foo": None},
+            "macros": {},
+            "labels": set("A"),
+            "extra_labels": set(),
+            "features": set(),
+            "components": set(),
+        }
+        mbed_lib_files = [
+            {
+                "path": Path("TARGET_A", "mbed_lib.json"),
+                "json_contents": {"name": "a", "target_overrides": {"*": {"target.features_add": ["RED"]}}},
+            },
+            {
+                "path": Path("subdir", "FEATURE_RED", "mbed_lib.json"),
+                "json_contents": {
+                    "name": "red",
+                    "target_overrides": {"A": {"target.features_add": ["BLUE"], "target.components_add": ["LEG"]}},
+                    "macros": ["RED_MACRO"],
+                },
+            },
+        ]
+        mbed_app_file = {
+            "path": Path("mbed_app.json"),
+            "json_contents": {
+                "target_overrides": {
+                    "*": {
+                        "target.foo": "bar",
+                        "target.labels_add": ["PICKLE"],
+                        "target.extra_labels_add": ["EXTRA_HOT"],
+                    }
+                }
+            },
+        }
+
+        with TemporaryDirectory() as directory:
+            _ = create_files(directory, mbed_lib_files)
+            created_mbed_app_file = create_files(directory, [mbed_app_file])[0]
+
+            _ = _assemble_config_from_sources_and_lib_files(
+                target, find_files("mbed_lib.json", Path(directory)), created_mbed_app_file
+            )
+
+            self.assertEqual(target["features"], {"RED", "BLUE"})
+            self.assertEqual(target["components"], {"LEG"})
+            self.assertEqual(target["extra_labels"], {"EXTRA_HOT"})
+            self.assertEqual(target["labels"], {"A", "PICKLE"})
```

Comment:
Should we check `target["macros"]` or `target["config"]` here as well?

Reply:
`target["config"]` is handled by the `Config` object and is checked in another test. I have added a check for `target["macros"]` here.